### ISSUE TITLE
Snackbar may not be visible long enough for some users

### ISSUE THREAD

**Issue by ssciolla** [tl-its-umich-edu/my-learning-analytics#806]
Thank you for contributing to this project!

- Make sure to search the [issues](https://github.com/tl-its-umich-edu/my-learning-analytics/issues) for duplicates first!

## Expected behavior (A clear and concise description of what you expected to happen) : 

When a user makes a change that triggers a Snackbar, the Snackbar should be visible long enough that the majority of users can read or perceive the message before it disappears.

## Describe the bug (Tell us what happens instead of the expected behavior) :

From the 10/28/19 accessibility review:

"Toasts/Snackbars may display for too short a time to be perceivable by some users. May also be unperceivable to others. ... Some users may not benefit from confirmation that their action was successful."

## Steps to reproduce :

1. On any of the views, choose an option from a dropdown menu or click a checkbox.
2. Then, click the checkbox labeled with "Remember my setting" that appears.
3. Observe the dark box (the Snackbar) that appears in the bottom left (and how long it is present) telling the user that the setting has been saved.

![Screen Shot 2019-11-27 at 9 42 15 AM](https://user-images.githubusercontent.com/35741256/69732990-9f60a180-10fa-11ea-9aa7-1c9ff055f9fd.png)

## Additional context (Add any other context about the problem here) :

The review suggests the following regarding the duration: "Fade in (or slide in) for 200-400ms for saccade time for the eye to refocus, then show for 50ms for each of the characters in the message."

The review also notes that the Snackbar's HTML element container should have the `role=alert` attribute-value pair.

Fixing this will likely involve making changes to the `UserSettingSnackbar` component or its usages in the views.

**Comment by jpyan** [comment]
Hey I'm trying to look into this bug yet I can't trigger any snackbars. Would it be possible that you give a more specific reproduce steps (screenshots are helpful too)? Thanks!

**Comment by ssciolla** [comment]
@jpyan, sure, I'll update the description with a few more details momentarily.

**Comment by justin0022** [comment]
> Hey I'm trying to look into this bug yet I can't trigger any snackbars. Would it be possible that you give a more specific reproduce steps (screenshots are helpful too)? Thanks!

@jpyan if you go to the GradeDistribution view, you should be able to trigger the snackbar by clicking on the `Show my grades` checkbox.

**Comment by jpyan** [comment]
Thanks, I can see now.

**Comment by justin0022** [comment]
You may find this line helpful to consider: https://github.com/tl-its-umich-edu/my-learning-analytics/blob/master/assets/src/components/UserSettingSnackbar.js#L35

**Comment by jpyan** [comment]
I'm confused on  "then show for 50ms for each of the characters in the message." part. So currently the default is set to 3000 ms. And say if I have a message saying "Setting saved successfully!" (which is already longer than some other messages may shown up), it has 27 characters and the time duration will be 1350ms according to the review suggestion. Wouldn't that be worse than before?  Would it make more sense to modify the 50ms measurement?

**Comment by jpyan** [comment]
Also, by adding "role=alert", I suppose you want the user not able to close a Snackbar. Then I think I can delete 
action={[
        <IconButton
          key='close'
          aria-label='close'
          color='inherit'
          onClick={() => setSavedSnackbarOpen(false)}
        >
          <CloseIcon />
        </IconButton>
      ]}
and  const [savedSnackbarOpen, setSavedSnackbarOpen] = useState(false). 
Will you approve or you still want keep those functions?

**Comment by ssciolla** [comment]
@jpyan, thank you for your work on this, I will think more about your questions and approach tomorrow and get back to you. I apologize for the delay.

**Comment by ssciolla** [comment]
Hi @jpyan, here are some thoughts on your questions:

1) Duration: I agree with your assessment that the times set out in the issue are confusing, in that they would result in shorter Snackbar duration. The [Material Design](https://material.io/components/snackbars/#behavior) spec suggests somewhere between four and 10 seconds, which makes more sense to me but is not very specific. I will follow up with our accessibility expert to see if he can clarify the instructions, and get back to you. For now, I would try what you suggested and bump the duration per character up to 150 ms; take a look at that and think about whether it seems more reasonable.

2) Can you clarify why you think we should remove the ability to close the Snackbar? Especially if the duration becomes longer, this is probably an option we want to continue to give users. If you saw somewhere that it was a best practice not to include `role=alert` and an action to close, please link to that resource.

Thanks again for working on this!

**Comment by ssciolla** [comment]
Regarding 1. above, I contacted our accessibility expert, and he things we stick to the Material Design guidelines. To that end, please try 150 ms and 200 ms per character and choose whichever you feel provides the most seamless experience (I would lean towards longer if you're unsure).

**Comment by jpyan** [comment]
Hey for your suggestion 2,  before I thought the only reason to include "role=alert" is because you don't want the user to be able to close the Snackbar. And since now you still want the close function, then I believe there is no need to add "role=alert" and no code is need to be deleted. My interpretation for "role=alert" is from below found in https://material-ui.com/zh/components/snackbars/
<img width="732" alt="Screen Shot 2019-12-06 at 6 37 17 PM" src="https://user-images.githubusercontent.com/35250601/70363697-6f7e7000-1857-11ea-9408-70eadc28293d.png">
Is there any other reason you want to include role=alert besides preventing user to close the Snackbar?

**Comment by ssciolla** [comment]
Hi @jpyan, this is interesting. It seems the reason for adding the `role` property to the message is to help accessibility technologies process the notification; see [W3C ARIA recommendation](https://www.w3.org/TR/wai-aria-1.1/#alert), linked to in the link you shared.

Here's the way I interpret the Material Design page for our purposes: we aren't requiring that users close the Snackbar themselves, it will simply fade away if they don't do anything. Thus, I don't see the harm in giving this extra close option. However, if we choose to keep that option, I think this next line becomes relevant: "If a Snackbar requires focus to close it, then content authors should use the role of alertdialog."

Okay, so I would do some testing: Check to see if the Snackbar has focus (i.e. you can interact with it using Tab and Enter/Return) when it appears. My own cursory examination suggests that it doesn't but that you can give it focus by hitting Tab a couple times. If it doesn't, see if changing the role to `alertdialog` gives it focus (i.e. to close it, you just have to press Enter/Return).

Looking at your PR history, I have one suggestion: You'll probably have to specify the `role` value using the `ContentProps` property of the `Snackbar` component. See how it's done in the example on the MD page you linked to. There are also API links to `Snackbar` and `SnackbarContent` there with more details.

I'm not an expert on this, so we're learning here together. Please let me know if you have questions or disagree with my interpretation of the MD/ARIA specs.

**Comment by jpyan** [comment]
Hi @ssciolla, thanks for your reply. In terms of testing: by hitting tab once / twice depending on the view, it does show that Snackbar has focus. And it can be closed by press Enter/Return. This is tested before I added the "alertdialog". 

Yet the make sure I follow the documentation, I still decided to add role =  alertdialog value using the ContentProps property since it is aligned with our aim here :"If a Snackbar requires focus to close it, then content authors should use the role of alertdialog." After added that and I tested again, and in terms of functionality, it still works the same as previous tests. 

So I don't see any other functionality problems here (The Snackbar appears long enough; you can use focus to close it or you can close it by clicking it or just let it disappear on its own) . If you agree with this, you can review my newest PR. Thanks!

**Comment by jennlove-um** [comment]
Testing passes:  The snackbar displays for enough time before fading on its own.  Also, I am able to tab to the snackbar and close it using the keyboard.  I can also use the mouse to close it manually.
